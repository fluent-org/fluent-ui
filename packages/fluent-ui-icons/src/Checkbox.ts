import createIcon, { JSX } from './utils/createIcon'

const jsx: JSX = [
  'svg',
  {
    viewBox: '0 0 2048 2048'
  },
  [
    'path',
    {
      d: 'M2048 0v2048H0V0h2048zm-128 128H128v1792h1792V128z'
    }
  ]
]

export default createIcon(jsx, 'Checkbox')
