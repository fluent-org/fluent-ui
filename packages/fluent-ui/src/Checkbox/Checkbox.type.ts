import * as React from 'react'
import * as PropTypes from 'prop-types'
import { StandardProps } from '..'

export interface CheckboxProps
  extends StandardProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement, 'onChange'> {
  checked?: boolean
  value?: string
  onChange?: (checked: boolean) => void
  disabled?: boolean
  children?: React.ReactNode
}

export const CheckboxPropTypes = {
  checked: PropTypes.bool,
  value: PropTypes.string,
  onChange: PropTypes.func,
  disabled: PropTypes.bool,
  children: PropTypes.node
}
