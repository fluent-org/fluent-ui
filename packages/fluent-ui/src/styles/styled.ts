import styled from 'styled-components'
import { th, variant, system, breakpoints } from '@xstyled/system'

export { styled, system, th, variant, breakpoints }

export * from 'styled-components'
export * from '@xstyled/system'
// export * from '@xstyled/prop-types'

export * from './stylize'
