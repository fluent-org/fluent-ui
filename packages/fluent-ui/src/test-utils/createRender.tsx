import * as React from 'react'
import * as PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

interface WrapperProps {
  children: React.ReactNode
}

const Wrapper: React.FC<WrapperProps> = ({ children }): React.ReactElement => (
  <ThemeProvider theme={{}}>{children}</ThemeProvider>
)

Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui: React.ReactElement, options: any) =>
  render(ui, { wrapper: Wrapper, ...options })

export { customRender as render }
