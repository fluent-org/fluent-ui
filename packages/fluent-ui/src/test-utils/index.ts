export * from '@testing-library/react'
// @ts-ignore
export * from './createRender'
