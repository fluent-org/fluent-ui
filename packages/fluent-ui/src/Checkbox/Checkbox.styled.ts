import { styled, css, th, variant } from '../styles/styled'

const checked = variant({
  prop: 'checked',
  default: false,
  variants: {
    true: css`
      color: ${th.color('white.default')};
      background-color: ${th.color('accent.default')};
      border-color: ${th.color('accent.default')};
    `,
    false: css`
      color: ${th.color('black.default')};
      background-color: transparent;
      border-color: ${th.color('primary.default')};
    `
  }
})

const disabled = variant({
  prop: 'disabled',
  default: false,
  variants: {
    true: css`
      border-color: ${th.color('primary.dark3')};
    `,
    false: css``
  }
})

export const StyledCheckboxWrapper = styled.div<{
  checked?: boolean
  disabled?: boolean
}>`
  width: 20px;
  height: 20px;
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  border: 2px solid;
  transition: ${th.transition('checkbox')};
  &:hover {
    border-color: ${th.color('primary.dark3')};
  }
  &:active {
    background-color: ${th.color('primary.dark3')};
  }

  ${checked}
  ${disabled}
`

export const StyledCheckbox = styled.input`
  width: 100%;
  height: 100%;
  opacity: 0;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  cursor: ${({ disabled }): string => (disabled ? 'not-allowed' : 'pointer')};
`
