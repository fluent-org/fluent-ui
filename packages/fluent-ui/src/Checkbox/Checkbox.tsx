import * as React from 'react'
import { StyledCheckboxWrapper, StyledCheckbox } from './Checkbox.styled'
import { Accept as AcceptIcon } from '@fluent-ui/icons'

interface CheckboxProps {
  checked?: boolean
  value?: string
  onChange?: (checked: boolean) => void
  disabled?: boolean
}

const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  (
    { checked, value, onChange, disabled }: CheckboxProps,
    ref
  ): React.ReactElement => {
    function handleChange(e: React.ChangeEvent<HTMLInputElement>): void {
      onChange && onChange(e.target.checked)
    }
    return (
      <StyledCheckboxWrapper checked={checked} disabled={disabled}>
        {checked && <AcceptIcon />}
        <StyledCheckbox
          ref={ref}
          type="checkbox"
          checked={checked}
          value={value}
          onChange={handleChange}
          disabled={disabled}
        />
      </StyledCheckboxWrapper>
    )
  }
)

Checkbox.displayName = 'FCheckbox'

export default Checkbox
