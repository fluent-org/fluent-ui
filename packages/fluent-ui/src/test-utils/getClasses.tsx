import * as React from 'react'
import { createUseStyles } from '@fluent-ui/styles'
import { Styles } from 'jss'
import { render } from './createRender'

export function getClasses<T extends string>(styles: Styles, name: string): Record<T, string> {
  const useStyles = createUseStyles<T>(styles, { name })
  let classes: Record<T, string>
  function Util(props: any): React.ReactElement {
    classes = useStyles(props)
    return <div />
  }
  render(<Util />)
  // @ts-ignore
  return classes
}
