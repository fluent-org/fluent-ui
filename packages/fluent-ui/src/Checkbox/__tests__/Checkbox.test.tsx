import * as React from 'react'
import { fireEvent } from '@testing-library/react'
import { getClasses, render } from '../../test-utils'
import '@testing-library/jest-dom/extend-expect'

import Checkbox, { name } from '../Checkbox'
import { styles } from '../Checkbox.styled'
import { CheckboxClassProps } from '../Checkbox.type'

const classes = getClasses<CheckboxClassProps>(styles, name)

describe('Checkbox', (): void => {
  test('should be support ref', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    const { container } = render(<Checkbox ref={ref} />)
    const { current: element } = ref
    const checkboxElement = container.querySelector(`input[type='checkbox']`) as HTMLInputElement
    expect(element).toEqual(checkboxElement)
  })

  test('should be controlled', async (): Promise<void> => {
    const ref = React.createRef<HTMLInputElement>()
    let checked = false
    function handleChange(c: boolean): void {
      checked = c
    }
    render(<Checkbox ref={ref} checked={checked} onChange={handleChange} />)
    fireEvent.click(ref.current!)
    expect(checked).toEqual(true)
  })

  test('should be render a disabled checkbox', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    const { container } = render(<Checkbox ref={ref} disabled={true} />)
    expect(container.firstChild).toHaveClass(classes.disabled)
    expect(ref.current!.disabled).toEqual(true)
  })
})
