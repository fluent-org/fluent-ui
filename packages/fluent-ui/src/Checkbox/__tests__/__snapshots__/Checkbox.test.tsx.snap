// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox basic 1`] = `
.c0 {
  width: 20px;
  height: 20px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid;
  -webkit-transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  color: #000000;
  background-color: transparent;
  border-color: #cccccc;
}

.c0:hover {
  border-color: #7a7a7a;
}

.c0:active {
  background-color: #7a7a7a;
}

.c1 {
  width: 100%;
  height: 100%;
  opacity: 0;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  cursor: pointer;
}

<div
  className="c0"
>
  <input
    className="c1"
    onChange={[Function]}
    type="checkbox"
  />
</div>
`;

exports[`Checkbox prop disabled 1`] = `
.c0 {
  width: 20px;
  height: 20px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid;
  -webkit-transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  color: #000000;
  background-color: transparent;
  border-color: #cccccc;
  border-color: #7a7a7a;
}

.c0:hover {
  border-color: #7a7a7a;
}

.c0:active {
  background-color: #7a7a7a;
}

.c1 {
  width: 100%;
  height: 100%;
  opacity: 0;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  cursor: not-allowed;
}

<div
  className="c0"
  disabled={true}
>
  <input
    className="c1"
    disabled={true}
    onChange={[Function]}
    type="checkbox"
  />
</div>
`;
